src/irmin-pack/unix/control_file.ml

```diff
@@ -139,11 +149,34 @@ module Serde = struct
         match route_version () with
         | Ok _ as x -> x
         | Error _ -> Error `Corrupted_control_file
+
+      (* Like [of_bin_string_unsafe] but without the checksum
+         verifications, only to be used for debuging purposes *)
+      let of_bin_string_unsafe s =
```

Comment:
Instead of adding a new function, I'd propose adding an optional argument to `of_bin_string` that allows skipping the checksum validation (maybe something like `?(skip_checksum_validation = false)`). That way we don't have to maintain _almost_ identical functions.

It'll be slightly annoying when you're reading the versioned payload since you'll have to match both `Valid` and `Invalid` cases, but not a big deal (and maybe there is a type trick of some sort) compared to the benefit of consolidating code.

Reply:
I agree with your point of view, with the pre-V5 functions, the checksum computation was done later, hence I had not needed the duplication. However after rebasing for the V5, I had simply picked this way, but wasnt happy with it.

If a flag is fine by you, I would be happy to change that (and indeed handle the Valid/Invalid later on)

Reply:
Is it possible to implement `of_bin_string` on top of `of_bin_string_unsafe`?

Reply:
I don't have a strong opinion other than I'd like to reduce the code duplication / clearly model what it is doing. I'm not quite sure what `unsafe` is supposed to mean in this context, which is another reason I proposed `skip_checksum_validation`. Maybe try both ways and see which feels better.

Reply:
Implementing `of_bin_string` on top looks good to me, but now that I think about it, maybe we actually would want the tool to tell us if the checksum has failed and give us something like
```json
{
  "is_checksum_valid": false,
  "actual_checksum": 0
  "payload": {"V5": {"foo": "..."}}
}
```
Instead of simply:
```json
{"V5": {"foo": "..."}}
```